File: advent-of-code/day05/python/one.py
from typing import List, Tuple

def parse_database(input_text: str) -> tuple[List[Tuple[int, int]], List[int]]:
    """
    Parse input into:
      - fresh_ranges: list of (start, end)
      - available_ids: list of ingredient IDs
    """
    lines = [line.rstrip("\n") for line in input_text.splitlines()]

    fresh_ranges: List[Tuple[int, int]] = []
    available_ids: List[int] = []

    reading_ranges = True
    for raw_line in lines:
        line = raw_line.strip()

        if line == "":
            reading_ranges = False
            continue

        if reading_ranges:
            start_str, end_str = line.split("-")
            range_start = int(start_str)
            range_end = int(end_str)
            fresh_ranges.append((range_start, range_end))
        else:
            available_ids.append(int(line))

    return fresh_ranges, available_ids


def merge_overlapping_ranges(ranges: List[Tuple[int, int]]) -> List[Tuple[int, int]]:
    """
    Merge overlapping or touching inclusive ranges.
    """
    if not ranges:
        return []

    ranges.sort(key=lambda r: (r[0], r[1]))

    merged: List[Tuple[int, int]] = []
    current_start, current_end = ranges[0]

    for next_start, next_end in ranges[1:]:
        if next_start <= current_end + 1:
            # Overlap or touch
            current_end = max(current_end, next_end)
        else:
            merged.append((current_start, current_end))
            current_start, current_end = next_start, next_end

    merged.append((current_start, current_end))
    return merged


def is_id_fresh(merged_ranges: List[Tuple[int, int]], ingredient_id: int) -> bool:
    """
    Hand-written binary search.
    Find the last range whose start <= ingredient_id.
    """
    left = 0
    right = len(merged_ranges) - 1
    candidate_index = -1

    while left <= right:
        mid = (left + right) // 2
        range_start, _ = merged_ranges[mid]

        if range_start <= ingredient_id:
            candidate_index = mid        # mid is a valid candidate
            left = mid + 1               # try to find a later one
        else:
            right = mid - 1

    if candidate_index == -1:
        return False

    range_start, range_end = merged_ranges[candidate_index]
    return range_start <= ingredient_id <= range_end


def count_fresh_available_ids(input_text: str) -> int:
    fresh_ranges, available_ids = parse_database(input_text)
    merged_ranges = merge_overlapping_ranges(fresh_ranges)

    fresh_count = 0
    for ingredient_id in available_ids:
        if is_id_fresh(merged_ranges, ingredient_id):
            fresh_count += 1

    return fresh_count


# ---- AoC-style stdin usage ----
if __name__ == "__main__":
    import sys
    input_data = sys.stdin.read()
    print(count_fresh_available_ids(input_data))

File: advent-of-code/day05/javascript/one.js
const fs = require("fs");
const path = require("path");

const inputPath = path.join(__dirname, "..", "input.txt");
const input = fs.readFileSync(inputPath, "utf-8").trim();

function countFreshIngredients(inputText) {
  const lines = inputText.split(/\r?\n/);

  // ---------- 1) Parse ranges and IDs ----------
  const rawRanges = [];
  const availableIds = [];

  let readingRanges = true;

  for (const rawLine of lines) {
    const line = rawLine.trim();

    // blank line switches from ranges section to IDs section
    if (line === "") {
      readingRanges = false;
      continue;
    }

    if (readingRanges) {
      // parse "start-end"
      const [startStr, endStr] = line.split("-");
      const start = Number(startStr);
      const end = Number(endStr);
      rawRanges.push([start, end]);
    } else {
      // parse single integer ID
      availableIds.push(Number(line));
    }
  }

  // Edge case: no ranges
  if (rawRanges.length === 0) return 0;

  // ---------- 2) Sort and merge overlapping ranges ----------
  rawRanges.sort((a, b) => a[0] - b[0] || a[1] - b[1]);

  const mergedRanges = [];
  for (const [start, end] of rawRanges) {
    if (mergedRanges.length === 0) {
      mergedRanges.push([start, end]);
      continue;
    }

    const lastIndex = mergedRanges.length - 1;
    const [lastStart, lastEnd] = mergedRanges[lastIndex];

    // If current range overlaps or touches the previous range:
    // currentStart <= lastEnd + 1  => merge them
    if (start <= lastEnd + 1) {
      mergedRanges[lastIndex] = [lastStart, Math.max(lastEnd, end)];
    } else {
      mergedRanges.push([start, end]);
    }
  }

  // ---------- 3) For each ID, check if it falls into any merged range ----------
  // Binary search helper: find the last range whose start <= id
  function isFresh(id) {
    let left = 0;
    let right = mergedRanges.length - 1;
    let candidateIndex = -1;

    while (left <= right) {
      const mid = Math.floor((left + right) / 2);
      const [rangeStart] = mergedRanges[mid];

      if (rangeStart <= id) {
        candidateIndex = mid; // mid could contain id, but we search further right
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    if (candidateIndex === -1) return false;

    const [start, end] = mergedRanges[candidateIndex];
    return start <= id && id <= end;
  }

  let freshCount = 0;
  for (const id of availableIds) {
    if (isFresh(id)) freshCount++;
  }

  return freshCount;
}

const result = countFreshIngredients(input);
console.log("Total fresh ingredient IDs:", result);